### PR TITLE
Add missing attributes for the DCAT Catalog Schema and Shape

### LINKED ISSUE #88: DCAT Schema and Shape is incomplete
The JSON schema file at https://github.com/International-Data-Spaces-Association/ids-specification/blob/main/catalog/message/schema/catalog-schema.json only contains a few attributes. Full DCAT compliance requires also the missing ones. Same for the Shape file.

### PR DESCRIPTION
**Pull request by sebbader-sap**
Closes #88

### CODE REVIEW

**Comment by sebbader-sap** [comment]
@juliapampus: PR is finally ready for review.

File: catalog/message/schema/catalog-schema.json [hunk 1]
```diff
@@ -1,39 +1,381 @@
...
+                            "const": "dcat:Catalog"
+                        },
+                        "dcat:dataset": {
+                            "type": "array",
+                            "items": {
+                                "$ref": "#/definitions/Dataset"
+                            },
+                            "minItems": 1
+                        },
+                        "dcat:record": {
```
Comment (juliapampus) [inline catalog/message/schema/catalog-schema.json:38]: Why is this introduced?
> This class is optional and not all catalogs will use it. It exists for catalogs where a distinction is made between metadata about a dataset or service and metadata about the entry in the catalog about the dataset or service.

Source: [dcat v3](https://www.w3.org/TR/vocab-dcat-3/#Class:Catalog_Record)

Comment (juliapampus) [reply catalog/message/schema/catalog-schema.json:38]: Same goes for `themeTaxonomy`, `hasPart`, and `homepage`. I would argue that not mentioning it here allows to still use the values. They are not mandatory anyway. It is inconsistent to name them but not the other attributes of `dcat:Catalog` (ref. superclasses [dcat:Dataset](https://www.w3.org/TR/vocab-dcat-3/#Class:Dataset) and [dcat:Resource](https://www.w3.org/TR/vocab-dcat-3/#Class:Resource)).

File: catalog/message/schema/catalog-schema.json [hunk 2]
```diff
@@ -1,39 +1,381 @@
...
+                        },
+                        "dcat:themeTaxonomy": {
+                            "$ref": "#/definitions/Reference"
+                        },
+                        "dct:hasPart": {
+                            "type": "array",
+                            "items": {
+                                "$ref": "#/definitions/ResourceChoice"
+                            },
+                            "minItems": 1
```
Comment (juliapampus) [inline catalog/message/schema/catalog-schema.json:56]: Why is this mandatory? See comment above

File: catalog/message/schema/catalog-schema.json
Comment (juliapampus) [inline catalog/message/schema/catalog-schema.json:1]: Maybe a stupid question: Is is necessary to describe parts of a json schema that should already exist (I assume someone this for dcat before?)? Can't they be referenced? Then, we would only need to define the parts that are dataspace protocol specific.

File: negotiation/message/schema/contract-schema.json
Comment (juliapampus) [inline negotiation/message/schema/contract-schema.json:1]: See comment for [catalog-schema.json](https://github.com/International-Data-Spaces-Association/ids-specification/pull/89/files#diff-4397065fe77e11fed25ef8e225e228001964265478fdc71935138bd331a4d5da)

File: catalog/message/schema/catalog-schema.json
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:1]: I am not aware of any officially provided SHACL or SHEX file, not even speaking about a JSON Schema one. Also, some projects have initial starters (including "@context", "@id", and "@type", not more) which are not sufficient for us.

File: negotiation/message/schema/contract-schema.json
Comment (sebbader-sap) [reply negotiation/message/schema/contract-schema.json:1]: Same for ODRL... RDF vocabularies (= the community creating them) really don't like validations...

File: catalog/message/schema/catalog-schema.json (hunk 1 above)
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:38]: > Why is this introduced?

I put it in because it's listed in the standard... If somebody uses them, it needs to be validated. We can of course decide to exclude it explicitly but as long as it's allowed, I propose to also have it in the schema.

Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:38]: > I would argue that not mentioning it here allows to still use the values.

That's correct, of course. However, we should think about closing the data objects by explicitly not allowing additional attributes... Or at least document our decision.

Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:38]: > it is inconsistent to name them but not the other attributes of dcat:Catalog

Did we miss any? If so, it was not intended.

File: catalog/message/schema/catalog-schema.json
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:1]: Would be great by the way to have them... In case somebody is better with Google than I am, please paste the link!

File: catalog/message/schema/catalog-schema.json [hunk 3]
```diff
@@ -1,39 +1,381 @@
...
+                            "$ref": "#/definitions/Reference"
+                        },
+                        "prov:wasGeneratedBy": {
+                            "$ref": "#/definitions/Reference"
                         }
+                    }
+                }
+            ]
+        },
+        "DataService": {
```
Comment (mkollenstart) [inline catalog/message/schema/catalog-schema.json:236]: DataService should also include Resource as it is a subclass of Resource in DCAT

File: catalog/message/schema/catalog-schema.json [hunk 4]
```diff
@@ -1,39 +1,381 @@
...
+                "dcat:servesDataset": {
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/definitions/Dataset"
+                    },
+                    "minItems": 1
+                }
+            }
+        },
+        "Distribution": {
```
Comment (mkollenstart) [inline catalog/message/schema/catalog-schema.json:254]: Shouldn't all classes also have a `@type` property (same goes for DataService, Dataset, Resource)

File: catalog/message/catalog.json [hunk 5]
```diff
@@ -1,12 +1,13 @@
...
   "dct:title": "Data Provider A Catalog",
   "dct:description": "A catalog of data items",
   "dct:publisher": "Data Provider A",
   "dcat:keyword": [
-    "traffic", "government"
+    "traffic",
+    "government"
   ],
   "dcat:service": [
     {
```
Comment (mkollenstart) [inline catalog/message/catalog.json:13]: I don't recognize dct:terms and dct:endpointUrl here, the dct:endpointUrl should be dcat:endpointURL I suppose. For dct:terms I can't really find what this means.

File: catalog/message/catalog.json [hunk 6]
```diff
@@ -25,31 +26,27 @@
       "dcat:keyword": [
```
Comment (mkollenstart) [inline catalog/message/catalog.json:26]: The dct:description and dcat:keyword should be a Multilanguage array instead of a plain string or array of plain strings.

File: catalog/message/catalog.json [hunk 7]
```diff
@@ -25,31 +26,27 @@
...
-              "duty": []
+              ]
             }
-          ],
-          "prohibition": [],
-          "obligation": []
+          ]
         }
       ],
       "dcat:distribution": [
```
Comment (mkollenstart) [inline catalog/message/catalog.json:52]: The dct:format should be a reference instead of a plain string. And the dcat:accessService should be a list of DataService instead of a plain string.

File: catalog/message/dataset.json [hunk 8]
```diff
@@ -1,44 +1,49 @@
...
     {
       "@type": "dcat:Distribution",
-      "dct:format": "dspace:s3+push",
-      "dcat:accessService": "urn:uuid:4aa2dcc8-4d2d-569e-d634-8394a8834d77"
+      "dct:format": {
+        "@id": "dspace:s3+push"
+      },
+      "dcat:accessService": [
+        {
+          "@id": "urn:uuid:4aa2dcc8-4d2d-569e-d634-8394a8834d77"
```
Comment (mkollenstart) [inline catalog/message/dataset.json:44]: Given another comment I added, the DataService should also include a `@type`, otherwise it should be a reference

**Review by mkollenstart — COMMENTED** [review]
I went through the catalog schemas and examples, and I really like the way it is moving. But there are some inconsistencies in there.

File: catalog/message/schema/catalog-schema.json [hunk 9]
```diff
@@ -1,39 +1,381 @@
...
+    "$id": "https://w3id.org/dspace/schemas/v0.8/CatalogSchema",
+    "definitions": {
+        "Catalog": {
+            "type": "object",
+            "allOf": [
+                {
+                    "$ref": "#/definitions/Dataset"
+                },
+                {
+                    "$ref": "#/definitions/Resource"
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:19]: Correct, I have updated it.

File: catalog/message/schema/catalog-schema.json [hunk 10]
```diff
@@ -1,39 +1,381 @@
...
+                    }
+                },
+                "dct:modified": {
+                    "type": "object",
+                    "properties": {
+                        "@id": {
+                            "type": "string"
+                        }
+                    }
+                },
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:117]: Valid point, will update the entries.

File: catalog/message/schema/catalog-schema.json [hunk 11]
```diff
@@ -1,39 +1,381 @@
...
+                "dct:modified": {
+                    "$ref": "#/definitions/time"
+                },
+                "dct:publisher": {
+                    "$ref": "#/definitions/Reference"
+                },
+                "dct:relation": {
+                    "$ref": "#/definitions/Reference"
+                },
+                "dct:title": {
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:184]: Exactly. `rdfs:Literal` is nasty to validate. Let's talk about it in our call today.

File: catalog/message/schema/catalog-schema.json (hunk 4 above)
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:254]: Not necessarily. As long as it's not at the beginning of a JSON, the serialiser usually knows the class through the property (JSON key). However, and that's then a real problem, in case several different classes (subclasses) are allowed, this approach won't work. On the other hand, I'd say that allowing different classes is a bad practice for our case anyway...

File: catalog/message/catalog.json (hunk 6 above)
Comment (sebbader-sap) [reply catalog/message/catalog.json:26]: Let's have a discussion about it. I am open for both.

File: catalog/message/schema/catalog-schema.json (hunk 4 above)
Comment (mkollenstart) [reply catalog/message/schema/catalog-schema.json:254]: I agree, but don't we already have this problem with `dct:hasPart` in Catalog which is either a Catalog, Dataset, or DataService?
So in that case we would need it for some objects and then it might be easier to include the `@type` everywhere to not require complex logic to see where to include it and where not.

File: catalog/message/schema/catalog-schema.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:184]: Ok, as I more or less did not continue the discussion, let me trigger it again here.

> dct:title is a rdfs:Literal so it could be multilanguage

100% correct from the perspective of DCAT and RDF. 100% horrible from the perspective of the server logic ;-)
My implicit proposal by suggesting `type: string` is to reduce the degrees of freedom coming with rdfs:Literal and only allow one variant. 

While for `dct:description` - as I feel - multi language tags are plausible, for `dct:title` I somehow feel different. Not based on facts, though. I simply can imagine that users do not pay too much attention on the language of a title / use the same title across different languages, while they actually do care for e.g. `dct:description`...

What do you think? Would the possibilities of `dct:title` a multilanguage strings create more value than it would create overhead?

File: catalog/message/schema/catalog-schema.json [hunk 12]
```diff
@@ -1,39 +1,381 @@
...
+                        },
+                        "dct:spatial": {
+                            "$ref": "#/definitions/Reference"
+                        },
+                        "dct:temporal": {
+                            "$ref": "#/definitions/Reference"
+                        },
+                        "prov:wasGeneratedBy": {
+                            "$ref": "#/definitions/Reference"
                         }
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:231]: Yes, that's a valid point: If even we in the group do not really know how to use it, or why it's needed, better not mention it. 

Proposal:
I removed these attributes from the schema file. Everyone can still use them nevertheless. If we learn that actually they are used significantly, we can add them again in the schema for a V1.1 as adding attributes generally is not seen as a breaking change.

File: catalog/message/schema/catalog-schema.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (mkollenstart) [reply catalog/message/schema/catalog-schema.json:184]: This links to my comment in https://github.com/International-Data-Spaces-Association/ids-specification/issues/132#issuecomment-1631994313

I think this can be solved inside the context with type coercion, with perhaps a default language set to `"en"`.

File: catalog/message/catalog.json (hunk 5 above)
Comment (sebbader-sap) [reply catalog/message/catalog.json:13]: > dct:endpointUrl should be dcat:endpointURL

Probably only a typo... At least I also wouldn't know what `dct:endpointUrl` is.

> For dct:terms I can't really find what this means.

Same. I replaced it with `dcat:endpointDescription`, which is not too far away.

File: catalog/message/schema/catalog-schema.json (hunk 4 above)
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:230]: Moved to #160 to track it independently.

**Comment by sebbader-sap** [comment]
Short update: JSON Schemas are now on the new level (not completely finished yet but done for the scope of this PR). Shapes will come soon.

**Comment by sebbader-sap** [comment]
@mkollenstart @juliapampus if you have time, please check the content again. Will ask for a review right when we have the shapes done.

File: negotiation/message/schema/contract-request-message-schema.json [hunk 13]
```diff
@@ -14,35 +14,29 @@
...
+					"$ref": "https://w3id.org/dspace/schemas/v0.8/contract-schema.json#/definitions/Offer"
 				},
 				"dspace:callbackAddress": {
-					"type" : "string",
-					"required": true
+					"type" : "string"
 				}
-			}
+			},
+			"required": [ "@context", "@type", "dspace:dataSet", "dspace:consumerPid", "dspace:providerPid", "dspace:offer", "dspace:callbackAddress" ]
```
Comment (juliapampus) [inline negotiation/message/schema/contract-request-message-schema.json:39]: `dspace:providerPid` may not be known when initiating a negotiation.

File: transfer/message/schema/transfer-error-schema.json [hunk 14]
```diff
@@ -14,35 +14,30 @@
...
 					"type": "array",
-					"required": false,
 					"minItems": 1,
 					"items": {
 
 					}
 				}
-			}
+			},
+			"required": [ "@context", "@type", "dspace:providerPid", "dspace:consumerPid", "dspace:code"]
```
Comment (juliapampus) [inline transfer/message/schema/transfer-error-schema.json:40]: `dspace:code` is not required

File: transfer/message/transfer-request-message.json [hunk 15]
```diff
@@ -4,6 +4,6 @@
   "dspace:consumerPid": "urn:uuid:32541fe6-c580-409e-85a8-8a9a32fbe833",
   "dspace:agreementId": "urn:uuid:e8dc8655-44c2-46ef-b701-4cffdc2faa44",
   "dct:format": "dspace:S3_AWS_PUSH",
-  "dspace:dataAddress": {},
+  "dspace:dataAddress": "",
```
Comment (juliapampus) [inline transfer/message/transfer-request-message.json:7]: Should be an object, no? #107

File: transfer/message/transfer-start-message.json [hunk 16]
```diff
@@ -3,5 +3,5 @@
   "@type": "dspace:TransferStartMessage",
   "dspace:providerPid": "urn:uuid:a343fcbf-99fc-4ce8-8e9b-148c97605aab",
   "dspace:consumerPid": "urn:uuid:32541fe6-c580-409e-85a8-8a9a32fbe833",
-  "dspace:dataAddress": {}
+  "dspace:dataAddress": ""
```
Comment (juliapampus) [inline transfer/message/transfer-start-message.json:6]: Should be an object, no? #107

File: transfer/message/transfer-process.json [hunk 17]
```diff
@@ -3,5 +3,5 @@
   "@type": "dspace:TransferProcess",
   "dspace:providerPid": "urn:uuid:a343fcbf-99fc-4ce8-8e9b-148c97605aab",
   "dspace:consumerPid": "urn:uuid:32541fe6-c580-409e-85a8-8a9a32fbe833",
-  "dspace:state": "dspace:REQUESTED"
-}
+  "dspace:transferState": "dspace:REQUESTED"
```
Comment (juliapampus) [inline transfer/message/transfer-process.json:6]: Should be `dspace:state` (pls align other files if something is inconsistent)

**Review by juliapampus — CHANGES_REQUESTED** [review]
Pls don't forget to update the PNGs when updating PUML files.

File: negotiation/message/schema/contract-negotiation-error-schema.json [hunk 18]
```diff
@@ -14,37 +14,40 @@
...
+						},
+						"@value": {
+							"type": "string"
+						}
+					},
+					"required": [ "@language", "@value"]
 				}
-			}
+			},
+			"required": [ "@context", "@type", "dspace:providerPid", "dspace:consumerPid", "dspace:code", "dspace:reason"]
```
Comment (juliapampus) [inline negotiation/message/schema/contract-negotiation-error-schema.json:50]: see transfer error, `dspace:code` is not required.

**Review by juliapampus — CHANGES_REQUESTED** [review]

File: catalog/message/schema/catalog-schema.json [hunk 19]
```diff
@@ -1,39 +1,338 @@
...
+                },
+                "dct:identifier": {
+                    "type": "string"
+                },
+                "dct:isReferencedBy": {
+                    "$ref": "#/definitions/Reference"
+                },
+                "dct:issued": {
+                    "$ref": "#/definitions/time"
+                },
```
Comment (mkollenstart) [inline catalog/message/schema/catalog-schema.json:155]: `dct:issued` is already in the context, so this should be a string

File: catalog/message/schema/catalog-schema.json [hunk 20]
```diff
@@ -1,39 +1,338 @@
...
+                "dcat:accessService": {
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/definitions/DataService"
+                    },
+                    "minItems": 1
+                },
+                "dcat:accessURL": {
+                    "$ref": "#/definitions/Reference"
+                },
```
Comment (mkollenstart) [inline catalog/message/schema/catalog-schema.json:223]: This isn't in the context yet, but might actually fit well in the context with `"@type": "@id"` given that it is always just an URL

File: catalog/message/schema/catalog-schema.json [hunk 21]
```diff
@@ -1,39 +1,338 @@
...
+                "dcat:byteSize": {
+                    "$ref": "#/definitions/decimal"
+                },
```
Comment (mkollenstart) [inline catalog/message/schema/catalog-schema.json:226]: `dcat:byteSize` is already in the context with a default type of `xsd:decimal`

File: catalog/message/schema/catalog-schema.json [hunk 22]
```diff
@@ -1,39 +1,338 @@
...
+                "dcat:compressFormat": {
+                    "$ref": "#/definitions/Reference"
+                },
+                "dcat:downloadURL": {
+                    "$ref": "#/definitions/Reference"
+                },
```
Comment (mkollenstart) [inline catalog/message/schema/catalog-schema.json:232]: Similar to `dcat:accessURL`

**Review by mkollenstart — CHANGES_REQUESTED** [review]
All comments are related to the JSON-LD context, perhaps for clarity it might be an idea to complete remove all the Reference, time, decimal subspecifications and fit them all into the context.

File: catalog/message/shape/catalog-error-shape.ttl [hunk 23]
```diff
@@ -13,7 +13,7 @@
 
 @prefix sh: <http://www.w3.org/ns/shacl#> .
 
-shapes:
+dspace_shapes:
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply catalog/message/shape/catalog-error-shape.ttl:16]: It's valid, yes. `shapes` as a term is somehow owned by the Shacl vocabulary, could bring us into trouble later.

File: catalog/message/shape/catalog-error-shape.ttl [hunk 24]
```diff
@@ -25,9 +25,28 @@ shapes:
...
+	sh:property [
+		sh:path dspace:code ;
+		sh:datatype xsd:string ;
+	] ;
+	sh:property [
+        sh:path dspace:reason ;
+        sh:node dspace:ReasonShape ;
+    ] .
+	
+dspace_shapes:ReasonShape
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply catalog/message/shape/catalog-error-shape.ttl:40]: Should have been a check for a `rdf:langString`... @mkollenstart did we ever discuss whether we enforce LangStrings over `xsd:string`?

[the remark this answers is not in the document]
Comment (sebbader-sap) [reply catalog/message/shape/catalog-error-shape.ttl:40]: Just noticed that, e.g., dct:description has both in our examples...

File: negotiation/message/contract-agreement-verification-message.json [hunk 25]
```diff
@@ -1,19 +1,16 @@
 {
   "@context":  "https://w3id.org/dspace/v0.8/context.json",
+  "@id": "urn:uuid:dcbf434c-eacf-4582-9a02-f8dd50120fd3",
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply negotiation/message/contract-agreement-verification-message.json:3]: must have been a left-over from the debugging.

File: transfer/message/shape/transfer-process-shape.ttl [hunk 26]
```diff
@@ -51,12 +51,18 @@ shapes:TransferProcessShape
...
+			dspace:STARTED
+			dspace:TERMINATED
+			dspace:COMPLETED
+			dspace:SUSPENDED
+		) ;
 		sh:severity sh:Violation ;
 		sh:maxCount 1 ;
 		sh:minCount 1 ;
-		sh:message "<https://raw.githubusercontent.com/International-Data-Spaces-Association/ids-specification/master/schemas/transfer-process-shape.ttl> (TransferProcessShape): An dspace:state property must point exactly one dspace:TransferState entity."@en ;
+		sh:message "<https://raw.githubusercontent.com/International-Data-Spaces-Association/ids-specification/master/schemas/transfer-process-shape.ttl> (TransferProcessShape): An dspace:state property must point exactly one dspace:State entity."@en ;
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply transfer/message/shape/transfer-process-shape.ttl:65]: The state (names) are overlapping... I feel like we can distinguish them accordingly also with the current names but not 100% sure yet. Nevertheless, I have adjusted the messages.

File: catalog/message/schema/catalog-schema.json (hunk 20 above)
Comment (sebbader-sap) [reply catalog/message/schema/catalog-schema.json:223]: Proposal: After we have finished this PR, let's block us one hour and together put more properties into the context file.

File: negotiation/message/contract-agreement-verification-message.json [hunk 27]
```diff
@@ -3,22 +3,9 @@
   "@type": "dspace:ContractAgreementVerificationMessage",
   "dspace:providerPid": "urn:uuid:a343fcbf-99fc-4ce8-8e9b-148c97605aab",
   "dspace:consumerPid": "urn:uuid:32541fe6-c580-409e-85a8-8a9a32fbe833",
-  "cred:credentialSubject": {
-    "@type": "dspace:ContractAgreementMessage",
-    "dspace:hash": "abcdefghijk12345"
+  "dspace:referencedContractAgreementMessage": {
```
Comment (juliapampus) [inline negotiation/message/contract-agreement-verification-message.json:6]: I would shorten this to `dspace:agreementMessage`, `dspace:message`, `dspace:hashedMessage` or  similar. We are never using the term "reference" for actual references.

File: negotiation/message/contract-agreement-message.json [hunk 28]
```diff
@@ -4,15 +4,21 @@
...
-    "dspace:providerId": "...",
+    "odrl:target": "urn:uuid:3dd1add4-4d2d-569e-d634-8394a8836d23",
     "dspace:timestamp": "2023-01-01T01:00:00Z",
-    "odrl:permission": [
-      {
-        ...
-      }
-    ]
-  }
+    "odrl:assigner": "urn:tsdshhs636378",
```
Comment (juliapampus) [inline negotiation/message/contract-agreement-message.json:11]: What about #92? We decided to use `dspace:consumerId` and `dspace:providerId` some time ago. 

(including @jimmarino in the conversation)

File: negotiation/message/diagram/contract-agreement-message.puml [hunk 29]
```diff
@@ -12,16 +12,14 @@ class "dspace:ContractAgreementMessage" {
...
 }
 
 class odrl:Agreement {
     @type : "odrl:Agreement"
     @id: String
-    odrl:target : xsd:anyURI
-    dspace:timestamp : xsd:dateTime
-    dspace:consumerId : String
-    dspace:providerId : String
+    ...
```
Comment (juliapampus) [inline negotiation/message/diagram/contract-agreement-message.puml:22]: `dspace:timestamp` was introduced with #82

(including @jimmarino in the conversation)

File: negotiation/message/diagram/contract-agreement-verification-message.puml [hunk 30]
```diff
@@ -11,35 +11,14 @@ class "dspace:ContractAgreementVerificationMessage" {
...
     dspace:consumerPid : String
-    cred:credentialSubject : dspace:ContractAgreementMessage
-    sec:proof : sec:Proof
+    dspace:referencedContractAgreementMessage : ReferencedMessage
 }
 
-class dspace:ContractAgreementMessage {
-    @type : "dspace:ContractAgreementMessage"
-    dspace:hash : String
+class ReferencedMessage {
```
Comment (juliapampus) [inline negotiation/message/diagram/contract-agreement-verification-message.puml:17]: Referring to my previous comment: use another term like `HashedMessage`? `ReferencedMessage` is maybe not that "meaningful" considering the properties and purpose...

File: negotiation/message/schema/contract-negotiation-error-schema.json [hunk 31]
```diff
@@ -14,37 +14,40 @@
...
+						},
+						"@value": {
+							"type": "string"
+						}
+					},
+					"required": [ "@language", "@value"]
 				}
-			}
+			},
+			"required": [ "@context", "@type", "dspace:providerPid", "dspace:consumerPid", "dspace:reason"]
```
Comment (juliapampus) [inline negotiation/message/schema/contract-negotiation-error-schema.json:50]: `dspace:reason` is not required (" The message can be sent at any state of a negotiation without providing an explanation.", [contract.negotiation.protocol.md](https://github.com/International-Data-Spaces-Association/ids-specification/blob/main/negotiation/contract.negotiation.protocol.md)). Same goes for the transfer error message.

**Review by juliapampus — CHANGES_REQUESTED** [review]

**Review by juliapampus — COMMENTED** [review]
Just a few minor comments. Rest looks fine 👍🏻

**Review by juliapampus — APPROVED** [review]

**Review by mkollenstart — APPROVED** [review]
